debian/docs/masala-cli.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "MASALA" "1" "September 2013" "" ""
.
.SH "NAME"
\fBmasala\fR \- P2P name resolution daemon
.
.SH "SYNOPSIS"
\fBmasala\fR [\-d] [\-q] [\-h hostname] [\-k password] [\-r realm] [\-p port] [\-x server] [\-y port] [\-u username]
.
.SH "DESCRIPTION"
\fBmasala\fR is a P2P name resolution daemon\. It resolves hostnames into IPv6 addresses by using a distributed hash table\. This DHT is compatible to the Kademlia DHT as used in Bittorrent clients like Transmission\.
.
.P
By default, masala sends the first packet to a multicast address\. So, for intranet use cases, you do not need a bootstrap server\. Just start masala on 2 nodes without any parameters\. It simply works\.
.
.P
If you would like to connect nodes around the globe, you may use the Bittorrent network\. Simply select a Bittorrent bootstrap server as seen in the example below\. Your client becomes a full member of the swarm and resolves info hashes to IPv6/port tuples\. The swarm on the other end does the same for you\. But in your case, the info hash represents a hostname instead of a torrent file\.
.
.P
A NSS module makes any hostname with \fI\.p2p\fR at the end transparently available to your Linux OS\.
.
.P
Masala uses a hostname cache for fast lookups within big swarms\. A queried hostname will be cached for 30 minutes\. Masala also lookups that queried hostname every 5 minutes on its own\. So, the cache stays up\-to\-date\. It stops 30 minutes after your last query for that hostname\.
.
.SH "FILES"
.
.TP
\fB/etc/nsswitch\.conf\fR
Masala gets attached to the \fIhosts\fR line\. The masala daemon is only used for hostnames with \fI\.p2p\fR at the end\.
.
.TP
\fB$HOME/\.masala\.conf\fR
This file gets written by the Masala daemon and contains the server port number\. The NSS and CLI program read this file to get knowledge about the current daemon port\.
.
.SH "OPTIONS"
.
.TP
\fB\-h\fR \fIhostname\fR
By default /etc/hostname is used to determine your hostname\. The SHA1 hash of the hostname becomes the announced info_hash\.
.
.TP
\fB\-n\fR \fInode id string\fR
By default a random node id gets computed on every startup\. For testing purposes it may be useful to keep the same node id all the time\. The above string is not used directly\. Instead its SHA1 hash is used\.
.
.TP
\fB\-k\fR \fIpassword\fR
Setting a password results in encrypting each packet with AES256\. The encrypted message is encapsulated in bencode too\. You effectively isolate your nodes from the rest of the world this way\. This method is not compatible to the Bittorrent network and only works between masala daemons\.
.
.TP
\fB\-r\fR \fIrealm\fR
Creating a realm affects the method how to compute the info hash\. It helps you to isolate your nodes and be part of a bigger swarm at the same time\. This is useful to handle duplicate hostnames\. So now, everybody may have his own https://owncloud\.p2p within his own realm for example\.
.
.TP
\fB\-p\fR \fIport\fR
Listen to this port (Default: UDP/6881)
.
.TP
\fB\-a\fR \fIport\fR
Kademlia also stores a port\. You can use masala\-cli to ask for that port and \-a to specify that port\. (Default: "6881")
.
.TP
\fB\-x\fR \fIserver\fR
Use server as a bootstrap server\. The server can be an IPv6 address, a FQHN like www\.example\.net or even a IPv6 multicast address\. (Default: ff0e::1)
.
.TP
\fB\-y\fR \fIport\fR
The bootstrap server will be addressed at this port\. (Default: UDP/6881)
.
.TP
\fB\-u\fR \fIusername\fR
When starting as root, use \-u to change the UID\.
.
.TP
\fB\-d\fR
Start as a daemon and run in background\. The output will be send to syslog\.
.
.TP
\fB\-v\fR
Verbose\.
.
.SH "INSTALLATION"
.
.IP "\(bu" 4
apt\-get install build\-essential
.
.IP "\(bu" 4
apt\-get install debhelper
.
.IP "\(bu" 4
apt\-get install git\-buildpackage
.
.IP "\(bu" 4
apt\-get install libpolarssl\-dev
.
.IP "\(bu" 4
git\-buildpackage
.
.IP "" 0
.
.SH "EXAMPLES"
Announce the hostname \fIowncloud\.p2p\fR within the realm TooManySecrets to an IPv6 Bittorrent bootstrap server\. That bootstrap server is maintained by the creator of the Transmission DHT stack:
.
.IP "" 4
.
.nf

$ masala \-h owncloud\.p2p \-r TooManySecrets \-x dht\.wifi\.pps\.jussieu\.fr
$ getent hosts owncloud\.p2p
$ masala\-cli owncloud\.p2p
.
.fi
.
.IP "" 0
.
.SH "BUGS"
Lack of IPv6 support by the providers\.
.
.SH "SEE ALSO"
nsswitch\.conf(5)


debian/changelog
masala (0.9) unstable; urgency=low

  * Use a sorted list for get_peers searches instead of bucket containers to save some bytes.
  * By default, be verbose in the console and quiet while running as a daemon.

 -- Aiko Barz <aiko@chroot.de>  Mon, 16 Sep 2013 23:25:00 +0100
masala (0.8) unstable; urgency=low

  * Limit the number of info_hashes, that can be stored
  * Limit the number of nodes, that can be stored per info_hash
  * Return NULL bytes as the token if the info_hash DB grows beyond all limits
  * Rotate VALUEs after each request for a fair IP distribution

 -- Aiko Barz <aiko@chroot.de>  Wed, 28 Aug 2013 22:22:00 +0100
masala (0.7) unstable; urgency=low

  * Query cache
  * Write hint file, that contains the server port. Needed by CLI+NSS

 -- Aiko Barz <aiko@chroot.de>  Sun, 11 Aug 2013 14:22:00 +0100
masala (0.6) unstable; urgency=low

  * Bittorrent compatibility

 -- Aiko Barz <aiko@chroot.de>  Mon, 08 Jul 2013 20:30:00 +0100
masala (0.5) unstable; urgency=low

  * Do not use lookup keys within FIND requests to save some bytes.

 -- Aiko Barz <aiko@chroot.de>  Tue, 02 Apr 2013 19:42:00 +0100
masala (0.4) unstable; urgency=low

  * Minor protocol changes
  * Logging changes

 -- Aiko Barz <aiko@chroot.de>  Fri, 15 Mar 2013 21:30:00 +0100
masala (0.3) unstable; urgency=low

  * Realm support

 -- Aiko Barz <aiko@chroot.de>  Tue, 12 Mar 2013 20:50:00 +0100
masala (0.2) unstable; urgency=low

  * Removed polarssl from the NSS module
  * Do the lookup hash computation within the daemon
  * Allow recursive requests only from localhost
  * Database expiry run

 -- Aiko Barz <aiko@chroot.de>  Thu, 07 Mar 2013 23:10:00 +0100
masala (0.1) unstable; urgency=low

  * masala: Initial release

 -- Aiko Barz <aiko@chroot.de>  Thu, 25 Oct 2012 22:49:00 +0100


src/masala-cli.c
/*
Copyright 2010 Aiko Barz

This file is part of masala/tumbleweed.

masala/tumbleweed is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

masala/tumbleweed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with masala/tumbleweed.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>

#include "masala-cli.h"
#include "malloc.h"
#include "file.h"
#include "ben.h"

int masala_lookup( const char *hostname, int size ) {
	IP sa;
	socklen_t salen = sizeof( IP );
	UCHAR buffer[BUF_SIZE];
	int sockfd = -1;
	int n = 0;
	struct timeval tv;
	UCHAR *p = NULL;
	int j = 0;
	IP sin;
	char ip_buf[INET6_ADDRSTRLEN+1];
	int port = 0;

	/* Load port from config file */
	port = masala_port();

	memset( &sa, '\0', salen );
	memset( buffer, '\0', BUF_SIZE );

	/* Setup UDP */
	sockfd = socket( AF_INET6, SOCK_DGRAM, 0 );
	if( sockfd < 0 ) {
		return 0;
	}

	/* Set receive timeout */
	tv.tv_sec = TIMEOUT;
	tv.tv_usec = 0;
	setsockopt( sockfd, SOL_SOCKET, SO_RCVTIMEO, (char *)&tv, sizeof( struct timeval ) );

	/* Setup IPv6 */
	sa.sin6_family = AF_INET6;
	sa.sin6_port = htons( port );
	if( !inet_pton( AF_INET6, "::1", &(sa.sin6_addr)) ) {
		return 0;
	}

	n = sendto( sockfd, hostname, size, 0, (struct sockaddr *)&sa, salen );
	if( n != size ) {
		return 0;
	}

	n = recvfrom( sockfd, buffer, BUF_OFF1, 0, (struct sockaddr *)&sa, &salen );
	if( n % 18 != 0 ) {
		return 0;
	}

	p = buffer;
	for( j=0; j<n; j+=18 ) {
		memset( &sin, '\0', sizeof(IP) );
		sin.sin6_family = AF_INET6;

		/* IP */
		memcpy( &sin.sin6_addr, p, 16 );
		p += 16;

		/* Port */
		memcpy( &sin.sin6_port, p, 2 );
		p += 2;

		printf("%s %i\n", inet_ntop( AF_INET6, &sin.sin6_addr, ip_buf,
			INET6_ADDRSTRLEN), ntohs(sin.sin6_port) );
	}

	return 1;
}

int masala_port( void ) {
	char filename[BUF_SIZE];
	int filesize = 0;
	UCHAR *fbuf = NULL;
	BEN *ben = NULL;
	BEN *port = NULL;
	int port_number = 0;

	if( ( getenv( "HOME")) == NULL ) {
		fail("Looking up $HOME failed");
	}

	snprintf( filename, BUF_SIZE, "%s/%s", getenv( "HOME" ), ".masala.conf" );

	if( !file_isreg( filename ) ) {
		fail("%s not found", filename );
	}

	filesize = file_size( filename );
	if( filesize <= 0 ) {
		fail("Config file %s is empty", filename );
	}

	if( ( fbuf = (UCHAR *) file_load( filename, 0, filesize ) ) == NULL ) {
		fail("Reading %s failed", filename );
	}

	if( !ben_validate( fbuf, filesize ) ) {
		fail( "Broken config in %s", filename );
	}

	/* Parse request */
	ben = ben_dec( fbuf, filesize );
	if( ben == NULL ) {
		fail( "Decoding %s failed", filename );
	}
	if( ben->t != BEN_DICT ) {
		fail( "%s does not contain a dictionary", filename );
	}

	port = ben_searchDictStr( ben, "port" );
	if( !ben_is_int( port ) ) {
		fail( "%s does not contain a port number" );
	}

	if( port->v.i < 1 || port->v.i > 65535 ) {
		fail( "%s contains an invalid port number" );
	}

	port_number = port->v.i;

	ben_free( ben );
	myfree( fbuf, "masala_port" );

	return port_number;
}

int main( int argc, char **argv ) {
	if( argc < 2 ) {
		return 1;
	}
	if( argv == NULL ) {
		return 1;
	}

	if( !str_isValidHostname( argv[1], strlen( argv[1] ) ) ) {
		fail( "%s is not a valid hostname", argv[1]);
	}

	if( ! masala_lookup( argv[1], strlen( argv[1] ) ) ) {
		fail( "Looking up %s failed", argv[1]);
	}
	
	return 0;
}
